Apps/Src/FaultState.c

```diff
@@ -3,117 +3,96 @@
-                nonrecoverableFaultHandler();
-            } else {
-                hall_fault_cnt++;
-                MotorController_Restart(); //re-initialize motor
-                return;
-            }
-        }
-
-        if(TritiumError & T_INIT_FAIL){
-            // Note: separate tripcnt from T_HALL_SENSOR_ERR
-            static uint8_t init_fault_cnt = 0;
-            if(init_fault_cnt>5){
-                nonrecoverableFaultHandler(); //we've failed to init the motor five times
-            } else {
-                init_fault_cnt++;
-                MotorController_Restart();
-                return;
-            }
-        }
-        return;
-
-        /**
-         * FAULTS NOT HANDLED :
-         * Low Voltage Lockout Error - Not really much we can do if we're not giving the controller enough voltage,
-         * and if we miss drive messages as a result, it will shut itself off.
-         * Temp error - Motor will throttle itself, and there's nothing we can do additional to cool it down
-         */
-    }
-    else if(FaultBitmap & FAULT_READBPS){ // Missed BPS Charge enable message
-        readBPS_ContactorHandler();
-        return;
-    }
-    else if(FaultBitmap & FAULT_UNREACH){ //unreachable code
+
+    printf("%s", currException.message);
+    if(currException.callback != NULL){
+        currException.callback();
+        } // Custom callback
+
+    switch (currException.prio)
+    {
+    case 0:
+        // Things could always be worse
+        // (Reserved for future cases of greater severity)
+    case 1:
         nonrecoverableFaultHandler();
+        break;
+    case 2:
+        // Priority level 2 only calls callback
+        break;
+    default:
+        break;
     }
-    else if(FaultBitmap & FAULT_DISPLAY){
-        static uint8_t disp_fault_cnt = 0;
-        if(disp_fault_cnt>3){
-            Display_Fault(OSErrLocBitmap, FaultBitmap);
-        } else {
-            disp_fault_cnt++;
-            Display_Reset();
-            return;
-        }
-    }
-    if(fromThread){//no recovering if fault state was entered outside of the fault thread
-        return;
-    }
-    while(1){;} 
+    
 }
 
 void Task_FaultState(void *p_arg) {
     OS_ERR err;
     CPU_TS ts;
 
-    FaultBitmap = FAULT_NONE;
-    OSErrLocBitmap = OS_NONE_LOC;
+    OSMutexCreate(&FaultState_Mutex, "Fault state mutex", &err);
+    OSSemCreate(&FaultState_Sem4, "Fault State Semaphore", 0, &err);
 
     // Block until fault is signaled by an assert
     while(1){
         OSSemPend(&FaultState_Sem4, 0, OS_OPT_PEND_BLOCKING, &ts, &err);
-        fromThread = true;
         EnterFaultState();
-        fromThread = false;
+        OSMutexPost(&FaultState_Mutex, OS_OPT_POST_NONE, &err);
+        assertOSError(OS_FAULT_STATE_LOC, err);  // We've finished handling the error and can post the mutex now
         OSTimeDlyHMSM(0,0,0,5,OS_OPT_TIME_HMSM_STRICT,&err);
```

Comment:
Why do we have a time delay in the fault thread? I would think we'd want to address faults as soon as possible

Reply:
If I had to guess (haven't read the rest) it's because the delay is a scheduling point allowing for the OS to take over and switch out of the fault thread. I agree with your point regarding the handling latency though, so maybe someone should look at all the scheduling point functions and see if there's a better solution / whether it's actually a problem.

Reply:
Got it, thanks!